src/services/rfqm_service.ts

```
@@ -0,0 +1,130 @@
+        const isSelling = sellAmount !== undefined;
+        const marketOperation = isSelling ? MarketOperation.Sell : MarketOperation.Buy;
+        const assetFillAmount = isSelling ? sellAmount! : buyAmount!;
+        const makerToken = buyToken;
+        const takerToken = sellToken;
+
+        // Fetch quotes
+        const opts = {
+            ...RFQM_DEFAULT_OPTS,
+            apiKey,
+            intentOnFilling: false, // TODO - safe to hardcode?
+            isIndicative: true,
+        };
+        const indicativeQuotes = await this._quoteRequestor.requestRfqmIndicativeQuotesAsync(
+            makerToken,
+            takerToken,
+            assetFillAmount,
+            marketOperation,
+            undefined,
+            opts,
+        );
+
+        // Filter out quotes that:
+        // - are for the wrong pair
+        // - cannot fill 100 % of the requested amount
+        // - expire in less than 3 minutes
+        //
+        // And sort by best price
+        const now = new BigNumber(Date.now());
+        const sortedQuotes = indicativeQuotes
+            .filter((q) => q.takerToken === takerToken && q.makerToken === makerToken)
+            .filter((q) => {
+                const requestedAmount = isSelling ? q.takerAmount : q.makerAmount;
+                return requestedAmount.gte(assetFillAmount);
+            })
+            // tslint:disable-next-line: custom-no-magic-numbers
+            .filter((q) => q.expiry.gte(now.plus(ONE_MINUTE_MS * 3)))
+            .sort((a, b) => {
+                // Want the most amount of maker tokens for each taker token
+                const aPrice = a.makerAmount.div(a.takerAmount);
+                const bPrice = b.makerAmount.div(b.takerAmount);
+                return bPrice.minus(aPrice).toNumber();
+            });
+
+        // No quotes found
+        if (sortedQuotes.length === 0) {
+            return Promise.resolve(EMPTY_QUOTE_RESPONSE);
```

Can we throw a `QuoteNotFoundError` or return undefined? I'd rather not return anything than return something that looks valid but is effectively invalid.

---

Yup, returning null